rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,143 @@
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Implement a flow that generates the signature payload for the image, then
+  calculates its signature and pushes it either to the registry where the image
+  is located, using the same credentials that were used to push the image, or
+  [to the registry specified in the `COSIGN_REPOSITORY` environment variable](#key-generation-and-storage).
+  This flow must happen after the image has been pushed to the registry.
+
+- If `cosign` fails to sign an image, the build should fail and output an error
+  message in the build log, so the operator can troubleshoot the issue. The
+  errors should also appear in any other places where error messages are
+  presented.
+
+- Whenever `kpack` signs an image using `cosign`, it should add these
+  annotations:
+  - Build number.
+  - Build time.
+
+- Signing with `cosign` should not affect any configurations that enable signing
+  using the Notary v1 mechanism. These two signing mechanisms should be able to
+  coexist.
+
+### Key generation and storage
+
+`cosign` can use file-based keys, key management systems or hardware keys to
+ingest private keys, as well as a keyless flow that requires integration with
+[Rekor](https://github.com/sigstore/rekor).
+
+`kpack` should not generate the keys used for signing and verification. The user
+should pass them in using one of the mechanisms supported by `cosign`.
+In this RFC, we suggest that the first mechanism implemented be the use of
+Kubernetes `Secret`s for sending the relevant data into the build pod, by means
+of attaching secrets to the service account used by `kpack`.
+These secrets should contain private key and private key passphrase.
+Optionally, the user can add a repository URL to the secret, to address the use
+case when pushing the signatures to a different registry is desired.
+
+`kpack` will be able to determine that the user wants to sign images using
+`cosign` by scanning the types of the secrets attached to the service account,
+which in this use case would be a custom type:
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cosign-key
+type: kpack.io/cosign-credentials
+stringData:
+  cosign-privatekey: <cosign-private-key>
+  passphrase: <cosign-private-key-passphrase>
+  # Optional repository URL, to be used as the environment variable
+  # COSIGN_REPOSITORY. If not set, pushes the signature collocated with
+  # the image.
+  cosign-repository-url: <repository-url>
```

Comment:
should this be `cosign-repository` to match the env var `COSIGN_REPOSITORY`

Reply:
In case we decide to go down the route of using the same names as generated by `cosign`, would it be fair to use `cosign.repository` as the key here, to maintain consistency, instead of using dashes as a separator?